src/pre-charge/pre_charge.cpp

```diff
@@ -0,0 +1,257 @@
+    case pre_charge::State::MC_OFF:
+        mcOffState();
+        break;
+    case pre_charge::State::ESTOPWAIT:
+        eStopState();
+        break;
+    case pre_charge::State::PRECHARGE:
+        prechargeState();
+        break;
+    case pre_charge::State::CONT_CLOSE:
+        contCloseState();
+        break;
+    case pre_charge::State::MC_ON:
+        mcOnState();
+        break;
+    case pre_charge::State::CONT_OPEN:
+        contOpenState();
+        break;
+    case pre_charge::State::DISCHARGE:
+        dischargeState();
+        break;
+    case pre_charge::State::FORWARD_DISABLE:
+        forwardDisableState();
+    default:
+        break;
+    }
+}
+
+void pre_charge::getSTO() {
+    batteryOneStatus = batteryOne.readPin();
+    batteryTwoStatus = batteryTwo.readPin();
+    eStopStatus = eStop.readPin();
+    //TODO: Add GFD
+
+    if (batteryOneStatus == IO::GPIO::State::HIGH && batteryTwoStatus == IO::GPIO::State::HIGH && eStopStatus == IO::GPIO::State::LOW) {
+        stoStatus = IO::GPIO::State::HIGH;
+    } else {
+        stoStatus = IO::GPIO::State::LOW;
+    }
+}
+
+void pre_charge::getMCKey() {
+    keyStatus = key.readPin();
+}
+
+void pre_charge::setPrecharge(int state) {
+    if (state) {
+        pc.writePin(IO::GPIO::State::HIGH);
+    } else {
+        pc.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::setDischarge(int state) {
+    if (state) {
+        dc.writePin(IO::GPIO::State::HIGH);
+    } else {
+        dc.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::setContactor(int state) {
+    if (state) {
+        cont.writePin(IO::GPIO::State::HIGH);
+    } else {
+        cont.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::setAPM(int state) {
+    if (state) {
+        apm.writePin(IO::GPIO::State::HIGH);
+    } else {
+        apm.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::setForward(int state) {
+    if (state) {
+        forward.writePin(IO::GPIO::State::HIGH);
+    } else {
+        forward.writePin(IO::GPIO::State::LOW);
+    }
+}
+
+void pre_charge::mcOffState() {
+    if (stoStatus == IO::GPIO::State::LOW) {
+        state = State::ESTOPWAIT;
+    } else if (stoStatus == IO::GPIO::State::HIGH && keyStatus == IO::GPIO::State::HIGH) {
+        state = State::PRECHARGE;
+    }
+    //else stay on MC_OFF
+}
+
+void pre_charge::mcOnState() {
+    if (stoStatus == IO::GPIO::State::LOW || keyStatus == IO::GPIO::State::LOW) {
+        state = State::FORWARD_DISABLE;
+    }
+    //else stay on MC_ON
+}
+
+void pre_charge::eStopState() {
+    if (stoStatus == IO::GPIO::State::HIGH) {
+        state = State::MC_OFF;
+    }
+    //else stay on E-Stop
+}
+
+void pre_charge::prechargeState() {
+    setPrecharge(1);
+    time::wait(5);//wait 5 tau
+    setPrecharge(0);
+    if (stoStatus == IO::GPIO::State::LOW || keyStatus == IO::GPIO::State::LOW) {
+        state = State::CONT_OPEN;
+    } else if (stoStatus == IO::GPIO::State::HIGH) {
```

Comment:
Is there a reason for this extra comparison? Won't the stoStatus always be either high or low?